### PR TITLE
Web/mobile controls (webrtc)

### PR DESCRIPTION
**Pull request by erikwessman**
This is just to get it into main, still have to fix:

1. Dont hardcode URL to the remote signalling server
2. Change the string in WebInputSource's LeftInputString() and RightInputString() to something else
3. Dont show the QR and lobby code at the start screen immediately, only after press start (important!1!1!)
4. Add a method NotifyChangePlayerColor(string clientId, string newColor) to webrtc manager that notifies web clients of their color (so we can show it in mobile UI 🤪)

### CODE REVIEW

**Comment by erikwessman** [comment]
Also look at this if you wanna know what tf going on
https://github.com/weave-game/signalling-server#readme

File: weave/Menus/LobbyDemo.tscn [hunk 1]
```diff
@@ -0,0 +1,49 @@
+[gd_scene load_steps=3 format=3 uid="uid://suha7hqfnwef"]
```
Comment (theowiik) [inline weave/Menus/LobbyDemo.tscn:1]: Re-added, is this used?

File: weave/Scripts/InputSources/Lobby.cs [hunk 2]
```diff
@@ -1,19 +1,39 @@
...
+
     private readonly IList<PlayerInfo> _playerInfos = new List<PlayerInfo>();
     public IReadOnlyList<PlayerInfo> PlayerInfos => _playerInfos.AsReadOnly();
 
     public int Count => PlayerInfos.Count;
 
     public bool Open { get; set; }
 
+    public string LobbyCode { get; set; }
+    private const int _lobbyCodeLength = 5;
```
Comment (theowiik) [inline weave/Scripts/InputSources/Lobby.cs:25]: namings

File: weave/Scripts/InputSources/Lobby.cs [hunk 3]
```diff
@@ -27,21 +47,36 @@ public void Join(IInputSource inputSource)
...
         UpdatePlayerInfos();
+        PlayerLeftListeners?.Invoke(inputSource);
     }
 
+    private static string GenerateLobbyCode(string allowedCharacters, int length)
+    {
+        var rnd = new Random();
+        var result = new char[length];
+        for (var i = 0; i < length; i++)
+            result[i] = allowedCharacters[rnd.Next(allowedCharacters.Length - 1)];
```
Comment (theowiik) [inline weave/Scripts/InputSources/Lobby.cs:65]: What if this code is taken 🤭 (probably not a concern for us, fix if you want to)

File: weave/Scripts/InputSources/WebInputSource.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
+using Godot;
+
+namespace Weave.InputSources;
+
+public sealed class WebInputSource : IInputSource
+{
+    public string Id;
```
Comment (theowiik) [inline weave/Scripts/InputSources/WebInputSource.cs:7]: To properties (just add { get; set; })

File: weave/Scripts/Multiplayer/Manager.cs [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+using Weave.InputSources;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Net.WebSockets;
+using System.Text;
+
+namespace Weave.Multiplayer;
+
+public partial class Manager : Node
```
Comment (theowiik) [inline weave/Scripts/Multiplayer/Manager.cs:14]: We should probably have a more descriptive name? Perhaps NetworkClientManager? or something like that

File: weave/Scripts/Multiplayer/Manager.cs [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+using SIPSorcery.Net;
+using Newtonsoft.Json;
+using Weave.InputSources;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Net.WebSockets;
+using System.Text;
+
+namespace Weave.Multiplayer;
```
Comment (theowiik) [inline weave/Scripts/Multiplayer/Manager.cs:12]: Im thinking if we should rename this folder/namespace to something like Networking, as it is multiplayer but the other input sources are also forms of multiplayer

File: weave/Scripts/Multiplayer/Manager.cs (hunk 5 above)
Comment (theowiik) [inline weave/Scripts/Multiplayer/Manager.cs:14]: Does it need to inherit from Node?

File: weave/Scripts/Multiplayer/Manager.cs [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        foreach (var clientId in _clientConnections.Keys)
+        {
+            var startMessage = new { type = "message", message = "start", clientId };
+            await SendWebSocketMessageAsync(JsonConvert.SerializeObject(startMessage));
+        }
+    }
+
+    public async void NotifyEndGameAsync()
+    {
+        foreach (var clientId in _clientConnections.Keys)
```
Comment (theowiik) [inline weave/Scripts/Multiplayer/Manager.cs:198]: Hmm looks good 🥸🥸🥸yes 🥸🥸🥸

File: weave/Scripts/Utils/GDScriptHelper.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
+using Godot;
+
+namespace Weave;
+
+public static class GDScriptHelper
+{
+    public static ImageTexture GenerateQRCodeFromString(string str)
```
Comment (theowiik) [inline weave/Scripts/Utils/GDScriptHelper.cs:7]: I mentioned this in the last PR, but perhaps ""removing"" the gdscript dependency with a IQrCodeGenerator

**Review by theowiik — CHANGES_REQUESTED** [review]
🌐 🌐 🌐 ☎️

**Comment by theowiik** [comment]
> Load solution '/home/runner/work/weave/weave/weave/Weave.sln' Done loading solution '/home/runner/work/weave/weave/weave/Weave.sln' Analyze solution '/home/runner/work/weave/weave/weave/Weave.sln' Analyze 'WeaveGame' 1/2 Scripts/MenuControllers/GameOverOverlay.cs(25,13): warning CS4014: Because this call is not awaited, execution of the current method continues before the call is completed. Consider applying the 'await' operator to the result of the call. Scripts/MenuControllers/StartScreen.cs(64,9): warning CS4014: Because this call is not awaited, execution of the current method continues before the call is completed. Consider applying the 'await' operator to the result of the call. Scripts/InputSources/IInputSource.cs(2,1): hidden CS8019: Unnecessary using directive. Scripts/InputSources/WebInputSource.cs(1,1): hidden CS8019: Unnecessary using directive. Scripts/MenuControllers/GameOverOverlay.cs(4,1): hidden CS8019: Unnecessary using directive. Analyze 'WeaveTests' 2/2 Done analyzing solution '/home/runner/work/weave/weave/weave/Weave.sln' in 00:07.98
> 
> 2 CS4014 Because this call is not awaited, execution of the current method continues before the call is completed 3 CS8019 Unnecessary using directive
> 
> 5 diagnostics found

Most of these can be fixed with: `roslynator fix .\Weave.sln --format`

**Comment by theowiik** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=weave-game_weave&pullRequest=78)
> 
> [Bug](https://sonarcloud.io/project/issues?id=weave-game_weave&pullRequest=78&resolved=false&types=BUG) [C](https://sonarcloud.io/project/issues?id=weave-game_weave&pullRequest=78&resolved=false&types=BUG) [3 Bugs](https://sonarcloud.io/project/issues?id=weave-game_weave&pullRequest=78&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=weave-game_weave&pullRequest=78&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=weave-game_weave&pullRequest=78&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=weave-game_weave&pullRequest=78&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=weave-game_weave&pullRequest=78&resolved=false&types=SECURITY_HOTSPOT) [E](https://sonarcloud.io/project/security_hotspots?id=weave-game_weave&pullRequest=78&resolved=false&types=SECURITY_HOTSPOT) [1 Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=weave-game_weave&pullRequest=78&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=weave-game_weave&pullRequest=78&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=weave-game_weave&pullRequest=78&resolved=false&types=CODE_SMELL) [11 Code Smells](https://sonarcloud.io/project/issues?id=weave-game_weave&pullRequest=78&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=weave-game_weave&pullRequest=78&metric=new_coverage&view=list) [0.0% Coverage](https://sonarcloud.io/component_measures?id=weave-game_weave&pullRequest=78&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=weave-game_weave&pullRequest=78&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=weave-game_weave&pullRequest=78&metric=new_duplicated_lines_density&view=list)
> 
> idea Catch issues before they fail your Quality Gate with our IDE extension sonarlint [SonarLint](https://www.sonarsource.com/products/sonarlint/features/connected-mode/?referrer=sonarcloud-welcome)

We can take a look at these at the next PR

**Review by elias-carlson — APPROVED** [review]
LGTM

**Review by elias-carlson — APPROVED** [review]
still LGTM

**Review by theowiik — APPROVED** [review]

**Pull request merged**